Avx512 extract most significant bits

anthonycanino opened the pull request:
PR addresses https://github.com/dotnet/runtime/issues/80820.

This adds a new type, `TYP_MASK` which represents Intel opmask `k` registers on xarch platform. Enough infrastructure was added so that `Vector512.ExtractMostSignificantBits` could be hardware accelerated on `AVX512` capable platforms, which requires the use of the k-mask registers. 

A future PR will implement mask registers in full, which depends upon `VectorMask` API proposal. At the moment, only a single mask register will ever be used, and is used as an internal register for the `ExtractMostSignificantBits` method. Follow up PRs will extend this to the `Vector512` comparison operators, which require a similar approach.

As an example, the following code...

```C#
ulong x = v1.ExtractMostSignificantBits(); // v1 is `Vector512<long>`
```

roughly lowers to

```asm
vmovups zmm0, zmmword ptr[rcx]
vpmovq2m k1, zmm0
kmovb rax, k1
```

anthonycanino committed:
Add `TYP_MASK` and `Vector512.ExtractMostSignificantBits`.

anthonycanino committed:
Rebase / rename error fix.

kunalspathak commented on src/coreclr/jit/vartype.h:
```diff
@@ -70,7 +70,17 @@ inline bool varTypeIsSIMD(T vt)
+}
+
+template <class T>
+inline bool varTypeIsMask(T vt)
```
This should be under `TARGET_AMD64`?

kunalspathak replied:
Here and at other places around `isMaskReg()`, etc.

tannergooding replied:
`TARGET_XARCH`, but yes we should try to keep the logic as non-impacting to Arm/Arm64.

tannergooding replied:
That's how we handle many other similar APIs. Looks like it was adjusted to do just that so I'll mark this thread as "resolved".

kunalspathak commented on src/coreclr/jit/lsra.cpp:
```diff
@@ -733,16 +735,15 @@ LinearScan::LinearScan(Compiler* theCompiler)
             availableRegs[i] = &availableDoubleRegs;
         }
-#endif
+        else if (thisType == TYP_MASK)
```
This should be under `#ifdef TARGET_AMD64`. That's why there are seeing arm64 TP regressions. Also, use `varTypeIsMask()`.

anthonycanino replied:
I think we want `#ifdef TARGET_XARCH` but yes, I will make the change.

kunalspathak commented on src/coreclr/jit/lsrabuild.cpp:
```diff
@@ -1393,6 +1393,17 @@ RefPosition* LinearScan::buildInternalFloatRegisterDefForNode(GenTree* tree, reg
 }
 
+#if defined(FEATURE_SIMD)
+RefPosition* LinearScan::buildInternalMaskRegisterDefForNode(GenTree* tree, regMaskTP internalCands)
```
same here.

kunalspathak commented on src/coreclr/jit/lsrabuild.cpp:
```diff
@@ -1393,6 +1393,17 @@ RefPosition* LinearScan::buildInternalFloatRegisterDefForNode(GenTree* tree, reg
+    // The candidate set should contain only float registers.
+    assert((internalCands & ~availableMaskRegs) == RBM_NONE);
+
+    RefPosition* defRefPosition = defineNewInternalTemp(tree, MaskRegisterType, internalCands);
```
```suggestion
    return defineNewInternalTemp(tree, MaskRegisterType, internalCands);
```

kunalspathak commented on src/coreclr/jit/hwintrinsiclistxarch.h:
```diff
@@ -847,6 +873,8 @@ HARDWARE_INTRINSIC(SSE2,            UCOMISD,
 HARDWARE_INTRINSIC(SSE41,           PTEST,                                      16,              2,     {INS_ptest,             INS_ptest,              INS_ptest,              INS_ptest,              INS_ptest,              INS_ptest,              INS_ptest,              INS_ptest,              INS_invalid,            INS_invalid},           HW_Category_SimpleSIMD,             HW_Flag_NoRMWSemant ...
 HARDWARE_INTRINSIC(AVX,             PTEST,                                       0,              2,     {INS_ptest,             INS_ptest,              INS_ptest,              INS_ptest,              INS_ptest,              INS_ptest,              INS_ptest,              INS_ptest,              INS_vtestps,            INS_vtestpd},           HW_Category_SimpleSIMD,             HW_Flag_NoRMWSemant ...
 
+HARDWARE_INTRINSIC(AVX512F,         MoveMaskSpecial,                            64,              1,     {INS_invalid,           INS_invalid,            INS_invalid,            INS_invalid,            INS_movd,               INS_invalid,            INS_invalid,            INS_invalid,            INS_invalid,            INS_invalid},           HW_Category_SIMDScalar,            HW_Flag_BaseTypeFrom ...
```
Wondering why do you need `MoveMaskSpecial`? Can't you handle it using `NI_Vector512_ExtractMostSignificantBits`?

tannergooding replied:
The other `MoveMask` instructions are modeled directly and we'll, long term, want `MoveMask` to be the same.

So having it as `MoveMaskSpecial` will presumably make it simpler to transition once the proper `AVX512F_MoveMask` support is added in the future.

kunalspathak commented on src/coreclr/jit/targetx86.h:
```diff
@@ -74,6 +74,10 @@
   #define REG_FP_LAST              REG_XMM7
+
+  #define REG_MASK_FIRST           REG_K0
+  #define REG_MASK_LAST            REG_K7
```
Since you will be just using `REG_K1`, why not just enable single register and do the rest in follow-up PR when you consume more of those?

anthonycanino replied:
I can make the change.

Is there a reason not to just enable all of them in this pass though? Is it for testing purposes?

tannergooding replied:
Presumably having it be more correct here when the initial support is being added will make it less error prone and overall simpler when its extended to be more than just a single selection, right?

kunalspathak replied:
That sounds fine to me. I just wanted to understand the reasoning behind adding all the registers even though they are not consumed.

kunalspathak reviewed (dismissed):
Made the 1st pass. I have reviewed just the portion that is not part of #80960

anthonycanino committed:
Review edits.

anthonycanino committed:
Formatting.

anthonycanino commented:
@dotnet/avx512-contrib can we get a rerun? Looks like some stuff was timed out?

anthonycanino commented:
Looks like everything is passing, I have made the requested changes if you would like to give it another pass.

kunalspathak commented on src/coreclr/jit/hwintrinsiccodegenxarch.cpp:
```diff
@@ -1666,6 +1671,49 @@ void CodeGen::genAvxOrAvx2Intrinsic(GenTreeHWIntrinsic* node)
+                    unreached();
+            }
+
+            assert(emitter::isMaskReg(maskReg));
```
For now, should we update the assert to be:
```suggestion
            assert(emitter::isMaskReg(maskReg) && (maskReg == K1));
```

kunalspathak commented:
Changes LGTM. Once #80960 is merged, let us do a final superpmi measurement before merging.

BruceForstall commented:
@anthonycanino Please resolve the conflicts.

anthonycanino committed:
Review edits.

anthonycanino commented:
I just pulled the changes onto a fresh branch based on main after its dependent PR was merged.

kunalspathak commented:
Seems like a TP hit, possibly because of added checks for K-instruction in emitter? Can you double check?

<img width="522" alt="image" src="https://user-images.githubusercontent.com/12488060/223227928-eb762270-a506-4024-9ce4-bfe02a05d507.png">


Also, I am surprised to see diffs in arm64.

There are few things that should be under `TARGET_XARCH` like `MaskRegisterType TYP_MASK`, `buildInternalMaskRegisterDefForNode` and few more in lsra. Likewise caller of `IsKInstruction` should wrap it inside `#ifdef TARGET_XARCH`.

anthonycanino commented:
> Seems like a TP hit, possibly because of added checks for K-instruction in emitter? Can you double check?
> 
> <img alt="image" width="522" src="https://user-images.githubusercontent.com/12488060/223227928-eb762270-a506-4024-9ce4-bfe02a05d507.png">
> 
> Also, I am surprised to see diffs in arm64.
> 
> There are few things that should be under `TARGET_XARCH` like `MaskRegisterType TYP_MASK`, `buildInternalMaskRegisterDefForNode` and few more in lsra. Likewise caller of `IsKInstruction` should wrap it inside `#ifdef TARGET_XARCH`.

Yes will do shortly.

BruceForstall commented on src/coreclr/jit/emitxarch.cpp:
```diff
@@ -9578,6 +9620,11 @@ const char* emitter::emitRegName(regNumber reg, emitAttr attr, bool varName)
 
+    if (isMaskReg(reg))
+    {
+        return emitKregName(reg);
```
I don't think you need `emitKregName` at all.
```suggestion
        return rn;
```

Any direct callers can just use `getRegName()` instead.

BruceForstall commented on src/coreclr/jit/hwintrinsiccodegenxarch.cpp:
```diff
@@ -1671,6 +1671,49 @@ void CodeGen::genAvxFamilyIntrinsic(GenTreeHWIntrinsic* node)
+                    unreached();
+            }
+
+            assert(emitter::isMaskReg(maskReg) && maskReg == REG_K1);
```
Should this have a `TODO` that the `maskReg == REG_K1` is temporary?

anthonycanino committed:
Review cleanup.

anthonycanino commented:
> Seems like a TP hit, possibly because of added checks for K-instruction in emitter? Can you double check?
> 
> <img alt="image" width="522" src="https://user-images.githubusercontent.com/12488060/223227928-eb762270-a506-4024-9ce4-bfe02a05d507.png">
> 
> Also, I am surprised to see diffs in arm64.
> 
> There are few things that should be under `TARGET_XARCH` like `MaskRegisterType TYP_MASK`, `buildInternalMaskRegisterDefForNode` and few more in lsra. Likewise caller of `IsKInstruction` should wrap it inside `#ifdef TARGET_XARCH`.

FYI, tpdff with no changes...


<details>
<summary>Overall (<span style="color:red">+0.32%</span> to <span style="color:red">+0.52%</span>)</summary>
<div style="margin-left:1em">

|Collection|PDIFF|
|---|--:|
|aspnet.run.windows.x64.checked.mch|<span style="color:red">+0.47%</span>|
|aspnet_block.run.windows.x64.checked.mch|<span style="color:red">+0.52%</span>|
|benchmarks.run.windows.x64.checked.mch|<span style="color:red">+0.34%</span>|
|coreclr_tests.run.windows.x64.checked.mch|<span style="color:red">+0.40%</span>|
|libraries.crossgen2.windows.x64.checked.mch|<span style="color:red">+0.43%</span>|
|libraries.pmi.windows.x64.checked.mch|<span style="color:red">+0.40%</span>|
|libraries_tests.pmi.windows.x64.checked.mch|<span style="color:red">+0.32%</span>|


</div></details>

<details>
<summary>MinOpts (<span style="color:red">+0.53%</span> to <span style="color:red">+1.18%</span>)</summary>
<div style="margin-left:1em">

|Collection|PDIFF|
|---|--:|
|aspnet.run.windows.x64.checked.mch|<span style="color:red">+0.98%</span>|
|aspnet_block.run.windows.x64.checked.mch|<span style="color:red">+0.90%</span>|
|benchmarks.run.windows.x64.checked.mch|<span style="color:red">+0.74%</span>|
|coreclr_tests.run.windows.x64.checked.mch|<span style="color:red">+0.60%</span>|
|libraries.crossgen2.windows.x64.checked.mch|<span style="color:red">+1.18%</span>|
|libraries.pmi.windows.x64.checked.mch|<span style="color:red">+0.82%</span>|
|libraries_tests.pmi.windows.x64.checked.mch|<span style="color:red">+0.53%</span>|


</div></details>

<details>
<summary>FullOpts (<span style="color:red">+0.23%</span> to <span style="color:red">+0.43%</span>)</summary>
<div style="margin-left:1em">

|Collection|PDIFF|
|---|--:|
|aspnet.run.windows.x64.checked.mch|<span style="color:red">+0.37%</span>|
|aspnet_block.run.windows.x64.checked.mch|<span style="color:red">+0.40%</span>|
|benchmarks.run.windows.x64.checked.mch|<span style="color:red">+0.34%</span>|
|coreclr_tests.run.windows.x64.checked.mch|<span style="color:red">+0.23%</span>|
|libraries.crossgen2.windows.x64.checked.mch|<span style="color:red">+0.43%</span>|
|libraries.pmi.windows.x64.checked.mch|<span style="color:red">+0.40%</span>|
|libraries_tests.pmi.windows.x64.checked.mch|<span style="color:red">+0.31%</span>|


</div></details>

<details>
<summary>Details</summary>
<div style="margin-left:1em">

All contexts:

|Collection|Base # instructions|Diff # instructions|PDIFF|
|---|--:|--:|--:|
|aspnet.run.windows.x64.checked.mch|137,122,434,940|137,770,274,733|<span style="color:red">+0.47%</span>|
|aspnet_block.run.windows.x64.checked.mch|28,196,864,957|28,342,998,178|<span style="color:red">+0.52%</span>|
|benchmarks.run.windows.x64.checked.mch|55,715,871,918|55,907,406,382|<span style="color:red">+0.34%</span>|
|coreclr_tests.run.windows.x64.checked.mch|820,707,835,085|823,964,475,471|<span style="color:red">+0.40%</span>|
|libraries.crossgen2.windows.x64.checked.mch|123,754,250,291|124,288,544,069|<span style="color:red">+0.43%</span>|
|libraries.pmi.windows.x64.checked.mch|234,167,937,468|235,100,838,193|<span style="color:red">+0.40%</span>|
|libraries_tests.pmi.windows.x64.checked.mch|506,101,763,851|507,705,699,987|<span style="color:red">+0.32%</span>|

MinOpts contexts:

|Collection|Base # instructions|Diff # instructions|PDIFF|
|---|--:|--:|--:|
|aspnet.run.windows.x64.checked.mch|22,723,762,098|22,946,475,537|<span style="color:red">+0.98%</span>|
|aspnet_block.run.windows.x64.checked.mch|6,648,732,841|6,708,641,851|<span style="color:red">+0.90%</span>|
|benchmarks.run.windows.x64.checked.mch|427,800,839|430,954,243|<span style="color:red">+0.74%</span>|
|coreclr_tests.run.windows.x64.checked.mch|367,123,809,127|369,344,596,817|<span style="color:red">+0.60%</span>|
|libraries.crossgen2.windows.x64.checked.mch|1,712,929|1,733,155|<span style="color:red">+1.18%</span>|
|libraries.pmi.windows.x64.checked.mch|1,337,999,031|1,348,970,414|<span style="color:red">+0.82%</span>|
|libraries_tests.pmi.windows.x64.checked.mch|5,373,609,215|5,402,007,210|<span style="color:red">+0.53%</span>|

FullOpts contexts:

|Collection|Base # instructions|Diff # instructions|PDIFF|
|---|--:|--:|--:|
|aspnet.run.windows.x64.checked.mch|114,398,672,842|114,823,799,196|<span style="color:red">+0.37%</span>|
|aspnet_block.run.windows.x64.checked.mch|21,548,132,116|21,634,356,327|<span style="color:red">+0.40%</span>|
|benchmarks.run.windows.x64.checked.mch|55,288,071,079|55,476,452,139|<span style="color:red">+0.34%</span>|
|coreclr_tests.run.windows.x64.checked.mch|453,584,025,958|454,619,878,654|<span style="color:red">+0.23%</span>|
|libraries.crossgen2.windows.x64.checked.mch|123,752,537,362|124,286,810,914|<span style="color:red">+0.43%</span>|
|libraries.pmi.windows.x64.checked.mch|232,829,938,437|233,751,867,779|<span style="color:red">+0.40%</span>|
|libraries_tests.pmi.windows.x64.checked.mch|500,728,154,636|502,303,692,777|<span style="color:red">+0.31%</span>|



</div></details>


tpdiff with most of `IsKInstruction` check commented out...


<details>
<summary>Overall (<span style="color:red">+0.17%</span> to <span style="color:red">+0.35%</span>)</summary>
<div style="margin-left:1em">

|Collection|PDIFF|
|---|--:|
|aspnet.run.windows.x64.checked.mch|<span style="color:red">+0.32%</span>|
|aspnet_block.run.windows.x64.checked.mch|<span style="color:red">+0.33%</span>|
|benchmarks.run.windows.x64.checked.mch|<span style="color:red">+0.26%</span>|
|coreclr_tests.run.windows.x64.checked.mch|<span style="color:red">+0.17%</span>|
|libraries.crossgen2.windows.x64.checked.mch|<span style="color:red">+0.35%</span>|
|libraries.pmi.windows.x64.checked.mch|<span style="color:red">+0.30%</span>|
|libraries_tests.pmi.windows.x64.checked.mch|<span style="color:red">+0.22%</span>|


</div></details>

<details>
<summary>MinOpts (<span style="color:red">+0.20%</span> to <span style="color:red">+0.86%</span>)</summary>
<div style="margin-left:1em">

|Collection|PDIFF|
|---|--:|
|aspnet.run.windows.x64.checked.mch|<span style="color:red">+0.49%</span>|
|aspnet_block.run.windows.x64.checked.mch|<span style="color:red">+0.42%</span>|
|benchmarks.run.windows.x64.checked.mch|<span style="color:red">+0.32%</span>|
|coreclr_tests.run.windows.x64.checked.mch|<span style="color:red">+0.22%</span>|
|libraries.crossgen2.windows.x64.checked.mch|<span style="color:red">+0.86%</span>|
|libraries.pmi.windows.x64.checked.mch|<span style="color:red">+0.39%</span>|
|libraries_tests.pmi.windows.x64.checked.mch|<span style="color:red">+0.20%</span>|


</div></details>

<details>
<summary>FullOpts (<span style="color:red">+0.14%</span> to <span style="color:red">+0.35%</span>)</summary>
<div style="margin-left:1em">

|Collection|PDIFF|
|---|--:|
|aspnet.run.windows.x64.checked.mch|<span style="color:red">+0.28%</span>|
|aspnet_block.run.windows.x64.checked.mch|<span style="color:red">+0.30%</span>|
|benchmarks.run.windows.x64.checked.mch|<span style="color:red">+0.26%</span>|
|coreclr_tests.run.windows.x64.checked.mch|<span style="color:red">+0.14%</span>|
|libraries.crossgen2.windows.x64.checked.mch|<span style="color:red">+0.35%</span>|
|libraries.pmi.windows.x64.checked.mch|<span style="color:red">+0.30%</span>|
|libraries_tests.pmi.windows.x64.checked.mch|<span style="color:red">+0.22%</span>|


</div></details>

<details>
<summary>Details</summary>
<div style="margin-left:1em">

All contexts:

|Collection|Base # instructions|Diff # instructions|PDIFF|
|---|--:|--:|--:|
|aspnet.run.windows.x64.checked.mch|137,122,327,070|137,555,517,439|<span style="color:red">+0.32%</span>|
|aspnet_block.run.windows.x64.checked.mch|28,196,936,039|28,290,027,003|<span style="color:red">+0.33%</span>|
|benchmarks.run.windows.x64.checked.mch|55,716,216,346|55,859,859,330|<span style="color:red">+0.26%</span>|
|coreclr_tests.run.windows.x64.checked.mch|820,707,397,894|822,143,099,586|<span style="color:red">+0.17%</span>|
|libraries.crossgen2.windows.x64.checked.mch|123,754,441,178|124,186,483,562|<span style="color:red">+0.35%</span>|
|libraries.pmi.windows.x64.checked.mch|234,169,440,084|234,865,933,264|<span style="color:red">+0.30%</span>|
|libraries_tests.pmi.windows.x64.checked.mch|506,106,782,715|507,212,774,247|<span style="color:red">+0.22%</span>|

MinOpts contexts:

|Collection|Base # instructions|Diff # instructions|PDIFF|
|---|--:|--:|--:|
|aspnet.run.windows.x64.checked.mch|22,723,762,125|22,835,457,571|<span style="color:red">+0.49%</span>|
|aspnet_block.run.windows.x64.checked.mch|6,648,733,217|6,676,854,359|<span style="color:red">+0.42%</span>|
|benchmarks.run.windows.x64.checked.mch|427,800,890|429,168,439|<span style="color:red">+0.32%</span>|
|coreclr_tests.run.windows.x64.checked.mch|367,123,738,248|367,915,727,518|<span style="color:red">+0.22%</span>|
|libraries.crossgen2.windows.x64.checked.mch|1,712,930|1,727,636|<span style="color:red">+0.86%</span>|
|libraries.pmi.windows.x64.checked.mch|1,337,998,813|1,343,217,279|<span style="color:red">+0.39%</span>|
|libraries_tests.pmi.windows.x64.checked.mch|5,373,613,051|5,384,456,794|<span style="color:red">+0.20%</span>|

FullOpts contexts:

|Collection|Base # instructions|Diff # instructions|PDIFF|
|---|--:|--:|--:|
|aspnet.run.windows.x64.checked.mch|114,398,564,945|114,720,059,868|<span style="color:red">+0.28%</span>|
|aspnet_block.run.windows.x64.checked.mch|21,548,202,822|21,613,172,644|<span style="color:red">+0.30%</span>|
|benchmarks.run.windows.x64.checked.mch|55,288,415,456|55,430,690,891|<span style="color:red">+0.26%</span>|
|coreclr_tests.run.windows.x64.checked.mch|453,583,659,646|454,227,372,068|<span style="color:red">+0.14%</span>|
|libraries.crossgen2.windows.x64.checked.mch|123,752,728,248|124,184,755,926|<span style="color:red">+0.35%</span>|
|libraries.pmi.windows.x64.checked.mch|232,831,441,271|233,522,715,985|<span style="color:red">+0.30%</span>|
|libraries_tests.pmi.windows.x64.checked.mch|500,733,169,664|501,828,317,453|<span style="color:red">+0.22%</span>|



</div></details>

Seems like it was having an impact. 

Let's see what the tpdiff looks like now --- most everything was behind proper ifdefs before, but `availableMaskRegs` instance variable was not, as well as `TYP_MASK` which would affect `arm` slightly. I've since moved.

anthonycanino committed:
Build fixes.

tannergooding commented on src/coreclr/jit/typelist.h:
```diff
@@ -64,6 +64,7 @@ DEF_TP(SIMD16   ,"simd16" , TYP_SIMD16,  TI_STRUCT,16,16, 16,   4,16, VTF_S|VTF_
 DEF_TP(SIMD32   ,"simd32" , TYP_SIMD32,  TI_STRUCT,32,32, 32,   8,16, VTF_S|VTF_VEC)
 DEF_TP(SIMD64   ,"simd64" , TYP_SIMD64,  TI_STRUCT,64,64, 64,  16,16, VTF_S|VTF_VEC)
 #endif // TARGET_XARCH
+DEF_TP(MASK     ,"mask"   , TYP_MASK,    TI_STRUCT,8, 8,  8,    2,8, VTF_S)
```
It is a struct and will eventually map to the `VectorMask<T>` type(s).

Under the hood we have just `k0-k7` which are 64-bits and while the actual hardware manuals don't differentiate further than `k#`, the .NET and C++ intrinsics do. -- .NET has `VectorMask#<T>` while C++ has `__mmask8/16/32/64`. So `VectorMask256<byte>` is equivalent to `__mmask16` since it has 16 entries.

Given that, is just `MASK` sufficient or do we need `MASK8/16/32/64`? If we have more than one, presumably we'd have this be `VTF_S|VTF_MSK`. -- I could see us going either way with there being different tradeoffs for each approach.

anthonycanino commented:
![image](https://user-images.githubusercontent.com/91913526/223505900-74116dfa-b215-4b8d-b97e-045393ae3640.png)

@kunalspathak  Looks better.

@BruceForstall  I mostly addressed your comments, with one caveat --- right now, `TYP_MASK` is defined if `FEATURE_SIMD` is present and for `XARCH` only, consistent with SIMD32 and SIMD64. This creates a bit of `ifdef` dependency.

BruceForstall commented:
@anthonycanino Where is the TP regression coming from?

anthonycanino commented:
> @anthonycanino Where is the TP regression coming from?

Investigating with `pin` now.

tannergooding commented on src/coreclr/jit/emitxarch.cpp:
```diff
@@ -34,8 +34,15 @@ bool emitter::IsSSEOrAVXInstruction(instruction ins)
+
 //------------------------------------------------------------------------
-// IsAvx512OrPriorInstruction: Is this an Avx512 or Avx or Sse instruction.
+// IsAvx512OrPriorInstruction: Is this an Avx512 or Avx or Sse or K (opmask) instruction.
+// Technically, K instructions would be considered under the VEX encoding umbrella, but due to
+// the instruction table encoding had to be pulled out with the rest of the `INST5` definitions.
```
Can you elaborate on this? Having two distinct ranges is going to make this check quite a bit more expensive.

tannergooding replied:
Looking, it looks like you're doing this to get access to `rr` which really exists to cover instructions like `inc`, `dec`, `push`, and `pop` which are "unary" and so are simply `inc reg` or `inc [addr]` -- It is definitely a confusing name and given that they encode the memory version using `mr`, we likely should've just used `rm` and removed the extra column from these...

Instructions like `kmovw` are instead `rm` or `mr` (`kmov reg, reg/mem` or `mov reg/mem, reg`) and so can be encoded using `INST4`

BruceForstall replied:
I believe there's a restriction that the list of instructions in the table must be INST5, then INST4, then INST3, etc. The fact that we have ranges (e.g., FIRST_SSE_INSTRUCTION, LAST_SSE_INSTRUCTION) seems like a bit of luck? Instead of ranges should we use the flags bits instead?

anthonycanino replied:
We would need to not use `INST5` then, since the way the tables are generated, they assume no gaps in the `INSTX` definitions, where `X` is a number. Correct?

tannergooding replied:
> I believe there's a restriction that the list of instructions in the table must be INST5, then INST4, then INST3, etc

Right. The point being that I believe we do not need (and actually shouldn't be using) INST5. We should be using INST3 for these instead and therefore these can be in the same general SIMD ranges so we got a simple check.

-- The reasoning being that these instructions are `ins mem, reg` (mr) or `ins reg, r/m` (rm) and you'll see that instructions like `mov` or `add` or SIMD instructions like `addps` or `movd` are encoded similarly. `rr` really just exists for the couple "unary" instructions which are simple `ins mem` or `ins reg`

anthonycanino replied:
I understand what you mean, just want to make sure about this case for `kmov` though.

We have

`kmov mem, k reg`
`kmov k reg, k reg / mem
`kmov k reg, gp reg`
`kmov gp reg, k reg`

So the `mr` and `rm` forms do not handle when we want to move to and from general purpose registers.

Are you saying we can still use INST3?

tannergooding replied:
> Are you saying we can still use INST3?

Right

What we have is:
* `r, r/m` - `VEX.L0.0F.W0 90 /r KMOVW k1, k2/m16`
* `m, r  ` - `VEX.L0.0F.W0 91 /r KMOVW m16, k1`
* `r, r  ` - `VEX.L0.0F.W0 92 /r KMOVW k1, r32`
* `r, r  ` - `VEX.L0.0F.W0 93 /r KMOVW r32, k1`

For opcode `91`, we'll statically know it's correct based on it being the `emitIns_C_R`, `emitIns_S_R`, or similar memory encoding paths. So it can just be the `mr` entry.

For opcode `90`, we'll likewise statically know its correct based on it being `emitIns_R_C`, `emitIns_R_S`, or similar memory encoding paths. So it can just be the `rm` default entry.

As I indicated above, `rm` is also typically used for `reg, reg`. However, in this case we have a complexity in that we have three different opcodes that all map to the `reg, reg` scenario. This means that in `emitIns_R_R` we have to do some kind of check on both `isMaskReg(dstReg)` and `isMaskReg(op1Reg)`. Since we must do the checks to determine which of the 3 opcodes (`90`, `92`, or `93`) are correct, using `rr` doesn't benefit us at all and we're instead effectively best off defaulting to `90` and then changing it to `92` or `93` based on the other checks.

tannergooding commented on src/coreclr/jit/emitxarch.cpp:
```diff
@@ -5971,6 +5998,15 @@ bool emitter::HasSideEffect(instruction ins, emitAttr size)
+        case INS_kmovd:
+        case INS_kmovq:
+        {
+            hasSideEffect = true;
```
This should be "false" for `kmovq` since its a 64-bit to 64-bit move and therefore doesn't zero extend, right?

tannergooding replied:
This is notably still unresolved

tannergooding commented on src/coreclr/jit/hwintrinsiccodegenxarch.cpp:
```diff
@@ -1671,6 +1671,51 @@ void CodeGen::genAvxFamilyIntrinsic(GenTreeHWIntrinsic* node)
+            instruction maskIns;
+            instruction kmovIns;
+
+            switch (baseType)
```
Might be worth a `TODO` covering that when we extend this to include other mask kinds, we'll need to update the handling.

That is, `TYP_BYTE` is `kmovq` for `VectorMask512<byte>`. But it is a `kmovd` for `VectorMask256<byte>` and a `kmovw` for `VectorMask128<byte>`

tannergooding commented on src/coreclr/jit/hwintrinsicxarch.cpp:
```diff
@@ -1249,6 +1249,22 @@ GenTree* Compiler::impSpecialIntrinsic(NamedIntrinsic        intrinsic,
 
+        case NI_Vector512_ExtractMostSignificantBits:
+        {
+            if (compOpportunisticallyDependsOn(InstructionSet_AVX512F) &&
+                compOpportunisticallyDependsOn(InstructionSet_AVX512BW) &&
+                compOpportunisticallyDependsOn(InstructionSet_AVX512DQ))
```
We've been using `IsBaselineVector512IsaSupported` to handle this, makes it a bit simpler and allows us to optimize it where possible.

tannergooding commented on src/coreclr/jit/lsra.h:
```diff
@@ -486,6 +487,12 @@ class RegRecord : public Referenceable
             registerType = FloatRegisterType;
         }
+#if defined(TARGET_XARCH) && defined(FEATURE_SIMD)
+        else if (emitter::isMaskReg(reg))
```
Given that `isGeneralRegister` is going to be the most common case, I think we probably want to factor this code a bit differently so that the most common path has the least checks.

Something like:
```
if (emitter::isGeneralRegister(reg))
{
    // The constructor defaults to IntRegisterType
    assert(registerType == IntRegisterType);
}
else if (emitter::isFloatReg(reg))
{
    registerType = FloatRegisterType;
}
else
{
    assert(emitter::isMaskReg(reg));
    registerType = MaskRegisterType;
}
```

anthonycanino committed:
Address throughput issues pertaining to `availableRegCount`.

BruceForstall commented:
[Diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=197241&view=ms.vss-build-web.run-extensions-tab)

# Throughput impact on Windows x64

The following shows the impact on throughput in terms of number of instructions executed inside the JIT. Negative percentages/lower numbers are better.

## linux arm64

<details>
<summary>Overall (<span style="color:green">-0.02%</span> to <span style="color:green">-0.01%</span>)</summary>
<div style="margin-left:1em">

|Collection|PDIFF|
|---|--:|
|benchmarks.run.linux.arm64.checked.mch|<span style="color:green">-0.01%</span>|
|coreclr_tests.run.linux.arm64.checked.mch|<span style="color:green">-0.01%</span>|
|libraries.crossgen2.linux.arm64.checked.mch|<span style="color:green">-0.02%</span>|
|libraries.pmi.linux.arm64.checked.mch|<span style="color:green">-0.01%</span>|
|libraries_tests.pmi.linux.arm64.checked.mch|<span style="color:green">-0.01%</span>|


</div></details>

<details>
<summary>MinOpts (<span style="color:green">-0.08%</span> to <span style="color:green">-0.01%</span>)</summary>
<div style="margin-left:1em">

|Collection|PDIFF|
|---|--:|
|benchmarks.run.linux.arm64.checked.mch|<span style="color:green">-0.06%</span>|
|coreclr_tests.run.linux.arm64.checked.mch|<span style="color:green">-0.01%</span>|
|libraries.crossgen2.linux.arm64.checked.mch|<span style="color:green">-0.08%</span>|
|libraries.pmi.linux.arm64.checked.mch|<span style="color:green">-0.03%</span>|
|libraries_tests.pmi.linux.arm64.checked.mch|<span style="color:green">-0.01%</span>|


</div></details>

<details>
<summary>FullOpts (<span style="color:green">-0.02%</span> to <span style="color:green">-0.00%</span>)</summary>
<div style="margin-left:1em">

|Collection|PDIFF|
|---|--:|
|benchmarks.run.linux.arm64.checked.mch|<span style="color:green">-0.01%</span>|
|libraries.crossgen2.linux.arm64.checked.mch|<span style="color:green">-0.02%</span>|
|libraries.pmi.linux.arm64.checked.mch|<span style="color:green">-0.01%</span>|
|libraries_tests.pmi.linux.arm64.checked.mch|<span style="color:green">-0.01%</span>|


</div></details>

<details>
<summary>Details</summary>
<div style="margin-left:1em">

All contexts:

|Collection|Base # instructions|Diff # instructions|PDIFF|
|---|--:|--:|--:|
|benchmarks.run.linux.arm64.checked.mch|71,829,708,758|71,825,058,259|<span style="color:green">-0.01%</span>|
|coreclr_tests.run.linux.arm64.checked.mch|1,120,118,011,248|1,120,045,044,886|<span style="color:green">-0.01%</span>|
|libraries.crossgen2.linux.arm64.checked.mch|126,407,564,245|126,385,333,940|<span style="color:green">-0.02%</span>|
|libraries.pmi.linux.arm64.checked.mch|239,547,008,704|239,517,121,722|<span style="color:green">-0.01%</span>|
|libraries_tests.pmi.linux.arm64.checked.mch|573,413,222,714|573,370,564,634|<span style="color:green">-0.01%</span>|

MinOpts contexts:

|Collection|Base # instructions|Diff # instructions|PDIFF|
|---|--:|--:|--:|
|benchmarks.run.linux.arm64.checked.mch|1,185,157,334|1,184,421,151|<span style="color:green">-0.06%</span>|
|coreclr_tests.run.linux.arm64.checked.mch|464,245,669,694|464,200,751,165|<span style="color:green">-0.01%</span>|
|libraries.crossgen2.linux.arm64.checked.mch|1,959,736|1,958,112|<span style="color:green">-0.08%</span>|
|libraries.pmi.linux.arm64.checked.mch|1,644,582,602|1,644,028,948|<span style="color:green">-0.03%</span>|
|libraries_tests.pmi.linux.arm64.checked.mch|7,106,434,127|7,105,485,508|<span style="color:green">-0.01%</span>|

FullOpts contexts:

|Collection|Base # instructions|Diff # instructions|PDIFF|
|---|--:|--:|--:|
|benchmarks.run.linux.arm64.checked.mch|70,644,551,424|70,640,637,108|<span style="color:green">-0.01%</span>|
|coreclr_tests.run.linux.arm64.checked.mch|655,872,341,554|655,844,293,721|<span style="color:green">-0.00%</span>|
|libraries.crossgen2.linux.arm64.checked.mch|126,405,604,509|126,383,375,828|<span style="color:green">-0.02%</span>|
|libraries.pmi.linux.arm64.checked.mch|237,902,426,102|237,873,092,774|<span style="color:green">-0.01%</span>|
|libraries_tests.pmi.linux.arm64.checked.mch|566,306,788,587|566,265,079,126|<span style="color:green">-0.01%</span>|



</div></details>


---

## osx arm64

<details>
<summary>Overall (<span style="color:green">-0.02%</span> to <span style="color:green">-0.01%</span>)</summary>
<div style="margin-left:1em">

|Collection|PDIFF|
|---|--:|
|benchmarks.run.osx.arm64.checked.mch|<span style="color:green">-0.01%</span>|
|coreclr_tests.run.osx.arm64.checked.mch|<span style="color:green">-0.01%</span>|
|libraries.crossgen2.osx.arm64.checked.mch|<span style="color:green">-0.02%</span>|
|libraries.pmi.osx.arm64.checked.mch|<span style="color:green">-0.01%</span>|
|libraries_tests.pmi.osx.arm64.checked.mch|<span style="color:green">-0.01%</span>|


</div></details>

<details>
<summary>MinOpts (<span style="color:green">-0.08%</span> to <span style="color:green">-0.01%</span>)</summary>
<div style="margin-left:1em">

|Collection|PDIFF|
|---|--:|
|benchmarks.run.osx.arm64.checked.mch|<span style="color:green">-0.03%</span>|
|coreclr_tests.run.osx.arm64.checked.mch|<span style="color:green">-0.01%</span>|
|libraries.crossgen2.osx.arm64.checked.mch|<span style="color:green">-0.08%</span>|
|libraries.pmi.osx.arm64.checked.mch|<span style="color:green">-0.03%</span>|
|libraries_tests.pmi.osx.arm64.checked.mch|<span style="color:green">-0.01%</span>|


</div></details>

<details>
<summary>FullOpts (<span style="color:green">-0.02%</span> to <span style="color:green">-0.00%</span>)</summary>
<div style="margin-left:1em">

|Collection|PDIFF|
|---|--:|
|benchmarks.run.osx.arm64.checked.mch|<span style="color:green">-0.01%</span>|
|libraries.crossgen2.osx.arm64.checked.mch|<span style="color:green">-0.02%</span>|
|libraries.pmi.osx.arm64.checked.mch|<span style="color:green">-0.01%</span>|
|libraries_tests.pmi.osx.arm64.checked.mch|<span style="color:green">-0.01%</span>|


</div></details>

<details>
<summary>Details</summary>
<div style="margin-left:1em">

All contexts:

|Collection|Base # instructions|Diff # instructions|PDIFF|
|---|--:|--:|--:|
|benchmarks.run.osx.arm64.checked.mch|38,164,974,081|38,162,491,846|<span style="color:green">-0.01%</span>|
|coreclr_tests.run.osx.arm64.checked.mch|859,418,196,995|859,356,913,472|<span style="color:green">-0.01%</span>|
|libraries.crossgen2.osx.arm64.checked.mch|133,045,192,209|133,021,837,042|<span style="color:green">-0.02%</span>|
|libraries.pmi.osx.arm64.checked.mch|248,363,814,282|248,331,667,102|<span style="color:green">-0.01%</span>|
|libraries_tests.pmi.osx.arm64.checked.mch|546,442,041,072|546,399,704,570|<span style="color:green">-0.01%</span>|

MinOpts contexts:

|Collection|Base # instructions|Diff # instructions|PDIFF|
|---|--:|--:|--:|
|benchmarks.run.osx.arm64.checked.mch|574,094,530|573,941,898|<span style="color:green">-0.03%</span>|
|coreclr_tests.run.osx.arm64.checked.mch|461,565,891,532|461,521,207,939|<span style="color:green">-0.01%</span>|
|libraries.crossgen2.osx.arm64.checked.mch|2,111,547|2,109,807|<span style="color:green">-0.08%</span>|
|libraries.pmi.osx.arm64.checked.mch|1,699,673,406|1,699,089,772|<span style="color:green">-0.03%</span>|
|libraries_tests.pmi.osx.arm64.checked.mch|7,623,536,291|7,622,404,286|<span style="color:green">-0.01%</span>|

FullOpts contexts:

|Collection|Base # instructions|Diff # instructions|PDIFF|
|---|--:|--:|--:|
|benchmarks.run.osx.arm64.checked.mch|37,590,879,551|37,588,549,948|<span style="color:green">-0.01%</span>|
|coreclr_tests.run.osx.arm64.checked.mch|397,852,305,463|397,835,705,533|<span style="color:green">-0.00%</span>|
|libraries.crossgen2.osx.arm64.checked.mch|133,043,080,662|133,019,727,235|<span style="color:green">-0.02%</span>|
|libraries.pmi.osx.arm64.checked.mch|246,664,140,876|246,632,577,330|<span style="color:green">-0.01%</span>|
|libraries_tests.pmi.osx.arm64.checked.mch|538,818,504,781|538,777,300,284|<span style="color:green">-0.01%</span>|



</div></details>


---

## linux x64

<details>
<summary>Overall (<span style="color:red">+0.13%</span> to <span style="color:red">+0.26%</span>)</summary>
<div style="margin-left:1em">

|Collection|PDIFF|
|---|--:|
|benchmarks.run.linux.x64.checked.mch|<span style="color:red">+0.13%</span>|
|coreclr_tests.run.linux.x64.checked.mch|<span style="color:red">+0.26%</span>|
|libraries.crossgen2.linux.x64.checked.mch|<span style="color:red">+0.16%</span>|
|libraries.pmi.linux.x64.checked.mch|<span style="color:red">+0.16%</span>|
|libraries_tests.pmi.linux.x64.checked.mch|<span style="color:red">+0.14%</span>|


</div></details>

<details>
<summary>MinOpts (<span style="color:red">+0.42%</span> to <span style="color:red">+0.67%</span>)</summary>
<div style="margin-left:1em">

|Collection|PDIFF|
|---|--:|
|benchmarks.run.linux.x64.checked.mch|<span style="color:red">+0.67%</span>|
|coreclr_tests.run.linux.x64.checked.mch|<span style="color:red">+0.45%</span>|
|libraries.crossgen2.linux.x64.checked.mch|<span style="color:red">+0.64%</span>|
|libraries.pmi.linux.x64.checked.mch|<span style="color:red">+0.58%</span>|
|libraries_tests.pmi.linux.x64.checked.mch|<span style="color:red">+0.42%</span>|


</div></details>

<details>
<summary>FullOpts (<span style="color:red">+0.12%</span> to <span style="color:red">+0.16%</span>)</summary>
<div style="margin-left:1em">

|Collection|PDIFF|
|---|--:|
|benchmarks.run.linux.x64.checked.mch|<span style="color:red">+0.13%</span>|
|coreclr_tests.run.linux.x64.checked.mch|<span style="color:red">+0.12%</span>|
|libraries.crossgen2.linux.x64.checked.mch|<span style="color:red">+0.16%</span>|
|libraries.pmi.linux.x64.checked.mch|<span style="color:red">+0.16%</span>|
|libraries_tests.pmi.linux.x64.checked.mch|<span style="color:red">+0.14%</span>|


</div></details>

<details>
<summary>Details</summary>
<div style="margin-left:1em">

All contexts:

|Collection|Base # instructions|Diff # instructions|PDIFF|
|---|--:|--:|--:|
|benchmarks.run.linux.x64.checked.mch|69,973,416,200|70,067,485,678|<span style="color:red">+0.13%</span>|
|coreclr_tests.run.linux.x64.checked.mch|870,208,033,751|872,470,252,016|<span style="color:red">+0.26%</span>|
|libraries.crossgen2.linux.x64.checked.mch|92,833,460,634|92,985,284,214|<span style="color:red">+0.16%</span>|
|libraries.pmi.linux.x64.checked.mch|223,281,109,753|223,646,234,128|<span style="color:red">+0.16%</span>|
|libraries_tests.pmi.linux.x64.checked.mch|512,114,946,866|512,853,456,023|<span style="color:red">+0.14%</span>|

MinOpts contexts:

|Collection|Base # instructions|Diff # instructions|PDIFF|
|---|--:|--:|--:|
|benchmarks.run.linux.x64.checked.mch|931,862,432|938,085,561|<span style="color:red">+0.67%</span>|
|coreclr_tests.run.linux.x64.checked.mch|372,511,404,092|374,175,448,655|<span style="color:red">+0.45%</span>|
|libraries.crossgen2.linux.x64.checked.mch|1,455,829|1,465,121|<span style="color:red">+0.64%</span>|
|libraries.pmi.linux.x64.checked.mch|1,323,192,173|1,330,840,704|<span style="color:red">+0.58%</span>|
|libraries_tests.pmi.linux.x64.checked.mch|6,268,867,376|6,294,950,342|<span style="color:red">+0.42%</span>|

FullOpts contexts:

|Collection|Base # instructions|Diff # instructions|PDIFF|
|---|--:|--:|--:|
|benchmarks.run.linux.x64.checked.mch|69,041,553,768|69,129,400,117|<span style="color:red">+0.13%</span>|
|coreclr_tests.run.linux.x64.checked.mch|497,696,629,659|498,294,803,361|<span style="color:red">+0.12%</span>|
|libraries.crossgen2.linux.x64.checked.mch|92,832,004,805|92,983,819,093|<span style="color:red">+0.16%</span>|
|libraries.pmi.linux.x64.checked.mch|221,957,917,580|222,315,393,424|<span style="color:red">+0.16%</span>|
|libraries_tests.pmi.linux.x64.checked.mch|505,846,079,490|506,558,505,681|<span style="color:red">+0.14%</span>|



</div></details>


---

## windows arm64

<details>
<summary>Overall (<span style="color:green">-0.02%</span> to <span style="color:green">-0.01%</span>)</summary>
<div style="margin-left:1em">

|Collection|PDIFF|
|---|--:|
|benchmarks.run.windows.arm64.checked.mch|<span style="color:green">-0.01%</span>|
|coreclr_tests.run.windows.arm64.checked.mch|<span style="color:green">-0.01%</span>|
|libraries.crossgen2.windows.arm64.checked.mch|<span style="color:green">-0.02%</span>|
|libraries.pmi.windows.arm64.checked.mch|<span style="color:green">-0.01%</span>|
|libraries_tests.pmi.windows.arm64.checked.mch|<span style="color:green">-0.01%</span>|


</div></details>

<details>
<summary>MinOpts (<span style="color:green">-0.08%</span> to <span style="color:green">-0.01%</span>)</summary>
<div style="margin-left:1em">

|Collection|PDIFF|
|---|--:|
|benchmarks.run.windows.arm64.checked.mch|<span style="color:green">-0.03%</span>|
|coreclr_tests.run.windows.arm64.checked.mch|<span style="color:green">-0.01%</span>|
|libraries.crossgen2.windows.arm64.checked.mch|<span style="color:green">-0.08%</span>|
|libraries.pmi.windows.arm64.checked.mch|<span style="color:green">-0.03%</span>|
|libraries_tests.pmi.windows.arm64.checked.mch|<span style="color:green">-0.02%</span>|


</div></details>

<details>
<summary>FullOpts (<span style="color:green">-0.02%</span> to <span style="color:green">-0.00%</span>)</summary>
<div style="margin-left:1em">

|Collection|PDIFF|
|---|--:|
|benchmarks.run.windows.arm64.checked.mch|<span style="color:green">-0.01%</span>|
|libraries.crossgen2.windows.arm64.checked.mch|<span style="color:green">-0.02%</span>|
|libraries.pmi.windows.arm64.checked.mch|<span style="color:green">-0.01%</span>|
|libraries_tests.pmi.windows.arm64.checked.mch|<span style="color:green">-0.01%</span>|


</div></details>

<details>
<summary>Details</summary>
<div style="margin-left:1em">

All contexts:

|Collection|Base # instructions|Diff # instructions|PDIFF|
|---|--:|--:|--:|
|benchmarks.run.windows.arm64.checked.mch|54,201,662,021|54,198,199,687|<span style="color:green">-0.01%</span>|
|coreclr_tests.run.windows.arm64.checked.mch|1,051,165,432,682|1,051,097,135,354|<span style="color:green">-0.01%</span>|
|libraries.crossgen2.windows.arm64.checked.mch|143,600,959,958|143,576,062,944|<span style="color:green">-0.02%</span>|
|libraries.pmi.windows.arm64.checked.mch|251,844,611,656|251,813,458,793|<span style="color:green">-0.01%</span>|
|libraries_tests.pmi.windows.arm64.checked.mch|551,090,992,404|551,050,520,240|<span style="color:green">-0.01%</span>|

MinOpts contexts:

|Collection|Base # instructions|Diff # instructions|PDIFF|
|---|--:|--:|--:|
|benchmarks.run.windows.arm64.checked.mch|545,947,019|545,795,218|<span style="color:green">-0.03%</span>|
|coreclr_tests.run.windows.arm64.checked.mch|465,525,194,605|465,479,805,506|<span style="color:green">-0.01%</span>|
|libraries.crossgen2.windows.arm64.checked.mch|2,118,887|2,117,147|<span style="color:green">-0.08%</span>|
|libraries.pmi.windows.arm64.checked.mch|1,733,086,668|1,732,517,304|<span style="color:green">-0.03%</span>|
|libraries_tests.pmi.windows.arm64.checked.mch|5,936,698,580|5,935,782,255|<span style="color:green">-0.02%</span>|

FullOpts contexts:

|Collection|Base # instructions|Diff # instructions|PDIFF|
|---|--:|--:|--:|
|benchmarks.run.windows.arm64.checked.mch|53,655,715,002|53,652,404,469|<span style="color:green">-0.01%</span>|
|coreclr_tests.run.windows.arm64.checked.mch|585,640,238,077|585,617,329,848|<span style="color:green">-0.00%</span>|
|libraries.crossgen2.windows.arm64.checked.mch|143,598,841,071|143,573,945,797|<span style="color:green">-0.02%</span>|
|libraries.pmi.windows.arm64.checked.mch|250,111,524,988|250,080,941,489|<span style="color:green">-0.01%</span>|
|libraries_tests.pmi.windows.arm64.checked.mch|545,154,293,824|545,114,737,985|<span style="color:green">-0.01%</span>|



</div></details>


---

## windows x64

<details>
<summary>Overall (<span style="color:red">+0.11%</span> to <span style="color:red">+0.27%</span>)</summary>
<div style="margin-left:1em">

|Collection|PDIFF|
|---|--:|
|aspnet.run.windows.x64.checked.mch|<span style="color:red">+0.22%</span>|
|aspnet_block.run.windows.x64.checked.mch|<span style="color:red">+0.25%</span>|
|benchmarks.run.windows.x64.checked.mch|<span style="color:red">+0.11%</span>|
|coreclr_tests.run.windows.x64.checked.mch|<span style="color:red">+0.27%</span>|
|libraries.crossgen2.windows.x64.checked.mch|<span style="color:red">+0.16%</span>|
|libraries.pmi.windows.x64.checked.mch|<span style="color:red">+0.15%</span>|
|libraries_tests.pmi.windows.x64.checked.mch|<span style="color:red">+0.14%</span>|


</div></details>

<details>
<summary>MinOpts (<span style="color:red">+0.43%</span> to <span style="color:red">+0.66%</span>)</summary>
<div style="margin-left:1em">

|Collection|PDIFF|
|---|--:|
|aspnet.run.windows.x64.checked.mch|<span style="color:red">+0.66%</span>|
|aspnet_block.run.windows.x64.checked.mch|<span style="color:red">+0.64%</span>|
|benchmarks.run.windows.x64.checked.mch|<span style="color:red">+0.58%</span>|
|coreclr_tests.run.windows.x64.checked.mch|<span style="color:red">+0.47%</span>|
|libraries.crossgen2.windows.x64.checked.mch|<span style="color:red">+0.66%</span>|
|libraries.pmi.windows.x64.checked.mch|<span style="color:red">+0.62%</span>|
|libraries_tests.pmi.windows.x64.checked.mch|<span style="color:red">+0.43%</span>|


</div></details>

<details>
<summary>FullOpts (<span style="color:red">+0.11%</span> to <span style="color:red">+0.16%</span>)</summary>
<div style="margin-left:1em">

|Collection|PDIFF|
|---|--:|
|aspnet.run.windows.x64.checked.mch|<span style="color:red">+0.12%</span>|
|aspnet_block.run.windows.x64.checked.mch|<span style="color:red">+0.13%</span>|
|benchmarks.run.windows.x64.checked.mch|<span style="color:red">+0.11%</span>|
|coreclr_tests.run.windows.x64.checked.mch|<span style="color:red">+0.11%</span>|
|libraries.crossgen2.windows.x64.checked.mch|<span style="color:red">+0.16%</span>|
|libraries.pmi.windows.x64.checked.mch|<span style="color:red">+0.15%</span>|
|libraries_tests.pmi.windows.x64.checked.mch|<span style="color:red">+0.13%</span>|


</div></details>

<details>
<summary>Details</summary>
<div style="margin-left:1em">

All contexts:

|Collection|Base # instructions|Diff # instructions|PDIFF|
|---|--:|--:|--:|
|aspnet.run.windows.x64.checked.mch|140,813,652,137|141,129,728,577|<span style="color:red">+0.22%</span>|
|aspnet_block.run.windows.x64.checked.mch|28,217,820,407|28,288,450,930|<span style="color:red">+0.25%</span>|
|benchmarks.run.windows.x64.checked.mch|55,793,824,417|55,856,927,078|<span style="color:red">+0.11%</span>|
|coreclr_tests.run.windows.x64.checked.mch|821,229,287,058|823,456,845,113|<span style="color:red">+0.27%</span>|
|libraries.crossgen2.windows.x64.checked.mch|123,855,428,806|124,049,862,818|<span style="color:red">+0.16%</span>|
|libraries.pmi.windows.x64.checked.mch|234,485,615,873|234,837,654,741|<span style="color:red">+0.15%</span>|
|libraries_tests.pmi.windows.x64.checked.mch|507,142,444,192|507,833,290,175|<span style="color:red">+0.14%</span>|

MinOpts contexts:

|Collection|Base # instructions|Diff # instructions|PDIFF|
|---|--:|--:|--:|
|aspnet.run.windows.x64.checked.mch|27,529,857,925|27,712,507,266|<span style="color:red">+0.66%</span>|
|aspnet_block.run.windows.x64.checked.mch|6,648,727,281|6,691,462,037|<span style="color:red">+0.64%</span>|
|benchmarks.run.windows.x64.checked.mch|427,797,145|430,288,887|<span style="color:red">+0.58%</span>|
|coreclr_tests.run.windows.x64.checked.mch|367,134,396,157|368,860,772,214|<span style="color:red">+0.47%</span>|
|libraries.crossgen2.windows.x64.checked.mch|1,712,897|1,724,206|<span style="color:red">+0.66%</span>|
|libraries.pmi.windows.x64.checked.mch|1,337,982,864|1,346,240,014|<span style="color:red">+0.62%</span>|
|libraries_tests.pmi.windows.x64.checked.mch|5,373,442,496|5,396,620,061|<span style="color:red">+0.43%</span>|

FullOpts contexts:

|Collection|Base # instructions|Diff # instructions|PDIFF|
|---|--:|--:|--:|
|aspnet.run.windows.x64.checked.mch|113,283,794,212|113,417,221,311|<span style="color:red">+0.12%</span>|
|aspnet_block.run.windows.x64.checked.mch|21,569,093,126|21,596,988,893|<span style="color:red">+0.13%</span>|
|benchmarks.run.windows.x64.checked.mch|55,366,027,272|55,426,638,191|<span style="color:red">+0.11%</span>|
|coreclr_tests.run.windows.x64.checked.mch|454,094,890,901|454,596,072,899|<span style="color:red">+0.11%</span>|
|libraries.crossgen2.windows.x64.checked.mch|123,853,715,909|124,048,138,612|<span style="color:red">+0.16%</span>|
|libraries.pmi.windows.x64.checked.mch|233,147,633,009|233,491,414,727|<span style="color:red">+0.15%</span>|
|libraries_tests.pmi.windows.x64.checked.mch|501,769,001,696|502,436,670,114|<span style="color:red">+0.13%</span>|



</div></details>


---


# Throughput impact on Windows x86

The following shows the impact on throughput in terms of number of instructions executed inside the JIT. Negative percentages/lower numbers are better.

## linux arm

<details>
<summary>Overall (<span style="color:green">-0.01%</span> to <span style="color:green">-0.00%</span>)</summary>
<div style="margin-left:1em">

|Collection|PDIFF|
|---|--:|
|libraries.crossgen2.linux.arm.checked.mch|<span style="color:green">-0.01%</span>|


</div></details>

<details>
<summary>MinOpts (<span style="color:green">-0.03%</span> to <span style="color:green">-0.00%</span>)</summary>
<div style="margin-left:1em">

|Collection|PDIFF|
|---|--:|
|benchmarks.run.linux.arm.checked.mch|<span style="color:green">-0.02%</span>|
|libraries.crossgen2.linux.arm.checked.mch|<span style="color:green">-0.03%</span>|
|libraries.pmi.linux.arm.checked.mch|<span style="color:green">-0.01%</span>|


</div></details>

<details>
<summary>FullOpts (<span style="color:green">-0.01%</span> to <span style="color:green">-0.00%</span>)</summary>
<div style="margin-left:1em">

|Collection|PDIFF|
|---|--:|
|libraries.crossgen2.linux.arm.checked.mch|<span style="color:green">-0.01%</span>|


</div></details>

<details>
<summary>Details</summary>
<div style="margin-left:1em">

All contexts:

|Collection|Base # instructions|Diff # instructions|PDIFF|
|---|--:|--:|--:|
|benchmarks.run.linux.arm.checked.mch|88,925,509,972|88,923,283,376|<span style="color:green">-0.00%</span>|
|coreclr_tests.run.linux.arm.checked.mch|1,077,827,852,065|1,077,805,328,402|<span style="color:green">-0.00%</span>|
|libraries.crossgen2.linux.arm.checked.mch|88,933,986,476|88,928,563,121|<span style="color:green">-0.01%</span>|
|libraries.pmi.linux.arm.checked.mch|243,437,368,111|243,426,162,137|<span style="color:green">-0.00%</span>|
|libraries_tests.pmi.linux.arm.checked.mch|508,087,880,991|508,072,032,434|<span style="color:green">-0.00%</span>|

MinOpts contexts:

|Collection|Base # instructions|Diff # instructions|PDIFF|
|---|--:|--:|--:|
|benchmarks.run.linux.arm.checked.mch|1,661,375,568|1,660,964,746|<span style="color:green">-0.02%</span>|
|coreclr_tests.run.linux.arm.checked.mch|443,043,133,437|443,030,285,490|<span style="color:green">-0.00%</span>|
|libraries.crossgen2.linux.arm.checked.mch|2,144,707|2,144,083|<span style="color:green">-0.03%</span>|
|libraries.pmi.linux.arm.checked.mch|1,656,826,419|1,656,621,067|<span style="color:green">-0.01%</span>|
|libraries_tests.pmi.linux.arm.checked.mch|7,321,475,376|7,321,117,500|<span style="color:green">-0.00%</span>|

FullOpts contexts:

|Collection|Base # instructions|Diff # instructions|PDIFF|
|---|--:|--:|--:|
|benchmarks.run.linux.arm.checked.mch|87,264,134,404|87,262,318,630|<span style="color:green">-0.00%</span>|
|coreclr_tests.run.linux.arm.checked.mch|634,784,718,628|634,775,042,912|<span style="color:green">-0.00%</span>|
|libraries.crossgen2.linux.arm.checked.mch|88,931,841,769|88,926,419,038|<span style="color:green">-0.01%</span>|
|libraries.pmi.linux.arm.checked.mch|241,780,541,692|241,769,541,070|<span style="color:green">-0.00%</span>|
|libraries_tests.pmi.linux.arm.checked.mch|500,766,405,615|500,750,914,934|<span style="color:green">-0.00%</span>|



</div></details>


---

## windows x86

<details>
<summary>Overall (<span style="color:red">+0.06%</span> to <span style="color:red">+0.10%</span>)</summary>
<div style="margin-left:1em">

|Collection|PDIFF|
|---|--:|
|benchmarks.run.windows.x86.checked.mch|<span style="color:red">+0.07%</span>|
|coreclr_tests.run.windows.x86.checked.mch|<span style="color:red">+0.10%</span>|
|libraries.crossgen2.windows.x86.checked.mch|<span style="color:red">+0.08%</span>|
|libraries.pmi.windows.x86.checked.mch|<span style="color:red">+0.08%</span>|
|libraries_tests.pmi.windows.x86.checked.mch|<span style="color:red">+0.06%</span>|


</div></details>

<details>
<summary>MinOpts (<span style="color:red">+0.15%</span> to <span style="color:red">+0.28%</span>)</summary>
<div style="margin-left:1em">

|Collection|PDIFF|
|---|--:|
|benchmarks.run.windows.x86.checked.mch|<span style="color:red">+0.23%</span>|
|coreclr_tests.run.windows.x86.checked.mch|<span style="color:red">+0.20%</span>|
|libraries.crossgen2.windows.x86.checked.mch|<span style="color:red">+0.28%</span>|
|libraries.pmi.windows.x86.checked.mch|<span style="color:red">+0.24%</span>|
|libraries_tests.pmi.windows.x86.checked.mch|<span style="color:red">+0.15%</span>|


</div></details>

<details>
<summary>FullOpts (<span style="color:red">+0.05%</span> to <span style="color:red">+0.08%</span>)</summary>
<div style="margin-left:1em">

|Collection|PDIFF|
|---|--:|
|benchmarks.run.windows.x86.checked.mch|<span style="color:red">+0.06%</span>|
|coreclr_tests.run.windows.x86.checked.mch|<span style="color:red">+0.05%</span>|
|libraries.crossgen2.windows.x86.checked.mch|<span style="color:red">+0.08%</span>|
|libraries.pmi.windows.x86.checked.mch|<span style="color:red">+0.08%</span>|
|libraries_tests.pmi.windows.x86.checked.mch|<span style="color:red">+0.06%</span>|


</div></details>

<details>
<summary>Details</summary>
<div style="margin-left:1em">

All contexts:

|Collection|Base # instructions|Diff # instructions|PDIFF|
|---|--:|--:|--:|
|benchmarks.run.windows.x86.checked.mch|63,484,806,822|63,526,339,468|<span style="color:red">+0.07%</span>|
|coreclr_tests.run.windows.x86.checked.mch|869,224,259,905|870,134,426,998|<span style="color:red">+0.10%</span>|
|libraries.crossgen2.windows.x86.checked.mch|129,233,207,301|129,339,015,895|<span style="color:red">+0.08%</span>|
|libraries.pmi.windows.x86.checked.mch|263,408,433,055|263,621,679,397|<span style="color:red">+0.08%</span>|
|libraries_tests.pmi.windows.x86.checked.mch|547,530,257,762|547,879,527,062|<span style="color:red">+0.06%</span>|

MinOpts contexts:

|Collection|Base # instructions|Diff # instructions|PDIFF|
|---|--:|--:|--:|
|benchmarks.run.windows.x86.checked.mch|413,217,394|414,153,173|<span style="color:red">+0.23%</span>|
|coreclr_tests.run.windows.x86.checked.mch|309,589,082,449|310,208,578,985|<span style="color:red">+0.20%</span>|
|libraries.crossgen2.windows.x86.checked.mch|2,103,307|2,109,257|<span style="color:red">+0.28%</span>|
|libraries.pmi.windows.x86.checked.mch|1,313,693,745|1,316,901,905|<span style="color:red">+0.24%</span>|
|libraries_tests.pmi.windows.x86.checked.mch|5,375,510,184|5,383,557,156|<span style="color:red">+0.15%</span>|

FullOpts contexts:

|Collection|Base # instructions|Diff # instructions|PDIFF|
|---|--:|--:|--:|
|benchmarks.run.windows.x86.checked.mch|63,071,589,428|63,112,186,295|<span style="color:red">+0.06%</span>|
|coreclr_tests.run.windows.x86.checked.mch|559,635,177,456|559,925,848,013|<span style="color:red">+0.05%</span>|
|libraries.crossgen2.windows.x86.checked.mch|129,231,103,994|129,336,906,638|<span style="color:red">+0.08%</span>|
|libraries.pmi.windows.x86.checked.mch|262,094,739,310|262,304,777,492|<span style="color:red">+0.08%</span>|
|libraries_tests.pmi.windows.x86.checked.mch|542,154,747,578|542,495,969,906|<span style="color:red">+0.06%</span>|



</div></details>


---

tannergooding commented on src/coreclr/jit/lsra.cpp:
```diff
@@ -698,6 +710,12 @@ LinearScan::LinearScan(Compiler* theCompiler)
     availableFloatRegs  = RBM_ALLFLOAT;
     availableDoubleRegs = RBM_ALLDOUBLE;
+#if defined(TARGET_XARCH)
+    if (compiler->DoJitStressEvexEncoding())
+    {
+        availableMaskRegs = RBM_ALLMASK;
+    }
```
Does this actually need to be conditioned? Presumably `availableMaskRegs` is completely unused if `DoJitStressEvexEncoding` is false so setting it as "available" wouldn't cause issues?

anthonycanino replied:
It is, so yes, makes sense to just let it default.

tannergooding commented on src/coreclr/jit/lsra.cpp:
```diff
@@ -737,6 +755,12 @@ LinearScan::LinearScan(Compiler* theCompiler)
             availableRegs[i] = &availableDoubleRegs;
         }
+#ifdef TARGET_XARCH
+        else if (thisType == TYP_MASK)
+        {
+            availableRegs[i] = &availableMaskRegs;
+        }
```
I think this is a general case like the other where integers are the most common, but they're at the bottom of the checked cases for the loop.

Presumably changing this to either a `switch` or to prioritize setting `availableIntRegs` as the first path would be beneficial so we don't have "one extra check" before hitting the path.

--------------------------

Actually, looking at this, this entire thing is just very unnecessarily expensive. We could track the classification in the `typelist.h` so we could do:
```cpp
#define DEF_TP(tn, nm, jitType, verType, sz, sze, asze, st, al, tf, ar) availableRegs[static_cast<int>(TYP_##tn)] = &ar;
#include "typelist.h"
#undef DEF_TP
    TYP_COUNT
```

This would involve no loops, no complex checks as to type kind/etc.

@kunalspathak, thoughts? Probably not that impactful overall, since its just part of the LSRA constructor; but still seems "meaningful" given we have around 25 `TYP_*` defines and this is doing 4-5 checks to initialize 17 of them and 3-4 checks to initialize 6 others.

tannergooding replied:
-- Not saying this needs to be handled in this PR, its something we can do as a follow up if its something we want to do.

kunalspathak replied:
I have updated #83109 to take care of it.

tannergooding commented on src/coreclr/jit/lsra.h:
```diff
@@ -1595,9 +1605,12 @@ class LinearScan : public LinearScanInterface
-    PhasedVar<regMaskTP>  availableFloatRegs;
-    PhasedVar<regMaskTP>  availableDoubleRegs;
+    PhasedVar<regMaskTP> availableIntRegs;
+    PhasedVar<regMaskTP> availableFloatRegs;
+    PhasedVar<regMaskTP> availableDoubleRegs;
```
@kunalspathak, it's not clear why we have `availableFloatRegs` and `availableDoubleRegs`. Could you elaborate a bit?

Given that `availableDoubleRegs` is used to account for both `double` and `TYP_SIMD` and that `float`/`double`/`simd` are all the "same registers", it's not immediately clear why we have the split.

Even on Arm32 where you have 1x double register taking up 2x float registers, it seems like it'd be simpler/easier to track this as a single "register file" and for double to simply take up 2 bits.

tannergooding replied:
-- Not something to handle in this PR, more making observations as this code is being touched in general and looking for potential opportunities to reduce the impact of changes like this, particularly given how impactful LSRA changes are to overall TP.

kunalspathak replied:
Honestly, I don't know historical reason behind having 2 separate variables for floats and double although we have `#define RBM_ALLDOUBLE            RBM_ALLFLOAT` in all targets except ARM. I will have to spend some time in understanding the implication if we unify them.

tannergooding commented on src/coreclr/vm/threadsuspend.cpp:
```diff
@@ -3050,7 +3050,7 @@ BOOL Thread::RedirectCurrentThreadAtHandledJITCase(PFN_REDIRECTTARGET pTgt, CONT
     ////////////////////////////////////////////////////
     // Now redirect the thread to the helper function
-
+   
```
Looks like this didn't get resolved "correctly".

The entire line was removed rather than just the whitespace change.

anthonycanino committed:
kmov RR refactor.

anthonycanino commented:
Does this look good now? The throughput numbers seem pretty good.

BruceForstall commented:
[Diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=199830&view=ms.vss-build-web.run-extensions-tab) -- looks good.

BruceForstall commented on src/coreclr/vm/threadsuspend.cpp:
```diff
@@ -3050,7 +3050,6 @@ BOOL Thread::RedirectCurrentThreadAtHandledJITCase(PFN_REDIRECTTARGET pTgt, CONT
 
     ////////////////////////////////////////////////////
     // Now redirect the thread to the helper function
-
```
The only change to this file is whitespace. Can you revert this change so `git` / GitHub don't see it as a change?

anthonycanino replied:
I removed it but let me see if I can revert it

BruceForstall commented on src/coreclr/jit/instrsxarch.h:
```diff
@@ -627,6 +627,18 @@ INST3(shrx,             "shrx",             IUM_WR, BAD_CODE,     BAD_CODE,
 
+INST3(FIRST_K_INSTRUCTION, "FIRST_K_INSTRUCTION", IUM_WR, BAD_CODE, BAD_CODE, BAD_CODE, INS_TT_NONE, INS_FLAGS_None )
+
+//    id                nm                  um      mr            mi            rm              tt              flags
```
nit: your comment header columns don't line up with the data columns below

BruceForstall reviewed (approved):
LGTM. Would like `threadsuspend.cpp` removed from the change.

Would like @tannergooding to also review.

anthonycanino committed:
Split kmov into kmov_msk and kmov_gpr.

anthonycanino committed:
Fix thread.

anthonycanino committed:
Review edits.

anthonycanino commented:
I think I've addressed all issues?

tannergooding commented:
Will get this reviewed a bit later tonight and should be able to get it merged at that time.

tannergooding commented on src/coreclr/jit/emitxarch.cpp:
```diff
@@ -34,8 +34,15 @@ bool emitter::IsSSEOrAVXInstruction(instruction ins)
 //------------------------------------------------------------------------
-// IsAvx512OrPriorInstruction: Is this an Avx512 or Avx or Sse instruction.
+// IsAvx512OrPriorInstruction: Is this an Avx512 or Avx or Sse or K (opmask) instruction.
+// Technically, K instructions would be considered under the VEX encoding umbrella, but due to
+// the instruction table encoding had to be pulled out with the rest of the `INST5` definitions.
```
Comment needs to be updated slightly since KInstructions are part of the INST3 group now.

tannergooding commented on src/coreclr/jit/emitxarch.cpp:
```diff
@@ -46,7 +53,7 @@ bool emitter::IsSSEOrAVXInstruction(instruction ins)
 {
     // TODO-XArch-AVX512: Fix check once AVX512 instructions are added.
-    return (ins >= INS_FIRST_SSE_INSTRUCTION) && (ins <= INS_LAST_AVX512_INSTRUCTION);
+    return ((ins >= INS_FIRST_SSE_INSTRUCTION) && (ins <= INS_LAST_AVX512_INSTRUCTION));
```
nit: 
```suggestion
    return (ins >= INS_FIRST_SSE_INSTRUCTION) && (ins <= INS_LAST_AVX512_INSTRUCTION);
```
